File: lib/gherkin/i18n_lexer.rb
require 'gherkin/i18n'
require 'gherkin/native'

module Gherkin
  I18nLexerNotFound = Class.new(LoadError)
  LexingError = Class.new(StandardError)

  # The main entry point to lexing Gherkin source.
  class I18nLexer
    native_impl('gherkin')

    LANGUAGE_PATTERN = /language\s*:\s*(.*)/ #:nodoc:
    attr_reader :i18n_language

    def initialize(listener, force_ruby=false)
      @listener = listener
      @force_ruby = force_ruby
    end

    def scan(source, uri)
      create_delegate(source).scan(source, uri)
    end

  private

    def create_delegate(source)
      @i18n_language = lang(source)
      @i18n_language.lexer(@listener, @force_ruby)
    end

    def lang(source)
      line_one = source.split(/\n/)[0]
      match = LANGUAGE_PATTERN.match(line_one)
      I18n.get(match ? match[1] : 'en')
    end

  end
end

File: java/src/main/java/gherkin/I18nLexer.java
package gherkin;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class I18nLexer implements Lexer {
    private static final Pattern LANGUAGE_PATTERN = Pattern.compile("language\\s*:\\s*(.*)");
    private final Listener listener;
    private I18n i18n;

    public I18nLexer(Listener listener) {
        this(listener, false);
    }

    public I18nLexer(Listener listener, boolean ignored) {
        this.listener = listener;
    }

    /**
     * @return the i18n language from the previous scanned source.
     */
    public I18n getI18nLanguage() {
        return i18n;
    }

    public void scan(CharSequence source) {
        createDelegate(source).scan(source);
    }

    private Lexer createDelegate(CharSequence source) {
        i18n = i18nLanguage(source);
        return i18n.lexer(listener);
    }

    private I18n i18nLanguage(CharSequence source) {
        String lineOne = source.toString().split("\\n")[0];
        Matcher matcher = LANGUAGE_PATTERN.matcher(lineOne);
        String key = "en";
        if (matcher.find()) {
            key = matcher.group(1);
        }
        return new I18n(key);
    }
}
